### PR TITLE
TryAdd, Keyed, and TryAddKeyed Attributes

### PR DESCRIPTION
**Pull request by ChaseDRedmon**
Add initial support for TryAdd, AddKeyed, and TryAddKeyed DI methods that are discussed in issues #5 and #6.

### CODE REVIEW

**Comment by ChaseDRedmon** [comment]
@patrickklaeren and @ristogod, here is an in-progress draft of what I'm working on to support the two other issues

File: src/AutoRegisterInject/Generator.cs [hunk 1]
```diff
@@ -97,7 +124,8 @@ private static AutoRegisteredClass GetAutoRegisteredClassDeclarations(GeneratorS
 
                     return new AutoRegisteredClass(typeName,
                         registrationType,
-                        registerAs);
+                        registerAs,
+                        string.Empty);
```
Comment (ChaseDRedmon) [inline src/AutoRegisterInject/Generator.cs]: In progress. Will add / support the constructor's new argument to define a key for Keyed methods and TryAddKeyed methods

File: src/AutoRegisterInject/AutoRegisteredClass.cs [hunk 2]
```diff
@@ -1,3 +1,3 @@
 ﻿namespace AutoRegisterInject;
 
-internal record AutoRegisteredClass(string ClassName, AutoRegistrationType RegistrationType, string[] Interfaces);
+internal record AutoRegisteredClass(string ClassName, AutoRegistrationType RegistrationType, string[] Interfaces, string? Key);
```
Comment (ChaseDRedmon) [inline src/AutoRegisterInject/AutoRegisteredClass.cs]: Adding the attribute constructor's key to this DTO here to support AddKeyed and TryAddKeyed methods

**Comment by ristogod** [comment]
@ChaseDRedmon Thanks. Looks good.

**Comment by ChaseDRedmon** [comment]
I'm going to open this PR as ready to review. 

I need to write the other unit test cases for testing the new TryRegister, TryRegisterKeyed, and RegisterKeyed methods under the Tests project, but I think we could start reviewing the PR initially @patrickklaeren.

I had one question on Ignored interfaces. I'm not sure if that's working correctly or what the functionality is supposed to be

**Comment by ChaseDRedmon** [comment]
Version bump would also need to be incremented to v1.4

File: src/AutoRegisterInject.Tests/GenerationTests.cs [hunk 3]
```diff
@@ -23,16 +23,73 @@ internal sealed class RegisterScopedAttribute : System.Attribute
 { 
     internal RegisterScopedAttribute(params System.Type[] onlyRegisterAs) { }
 }
+
+[System.AttributeUsage(System.AttributeTargets.Class, Inherited = false, AllowMultiple = false)]
+internal sealed class RegisterKeyedScopedAttribute : System.Attribute
+{ 
+    internal RegisterKeyedScopedAttribute(string ServiceKey, params System.Type[] onlyRegisterAs) { }
```
Comment (patrickklaeren) [inline src/AutoRegisterInject.Tests/GenerationTests.cs]:
```suggestion
    internal RegisterKeyedScopedAttribute(string serviceKey, params System.Type[] onlyRegisterAs) { }
```

File: src/AutoRegisterInject.Tests/GenerationTests.cs [hunk 4]
```diff
@@ -23,16 +23,73 @@ internal sealed class RegisterScopedAttribute : System.Attribute
...
+[System.AttributeUsage(System.AttributeTargets.Class, Inherited = false, AllowMultiple = false)]
+internal sealed class TryRegisterScopedAttribute : System.Attribute
+{ 
+    internal TryRegisterScopedAttribute(params System.Type[] onlyRegisterAs) { }
+}
+
+[System.AttributeUsage(System.AttributeTargets.Class, Inherited = false, AllowMultiple = false)]
+internal sealed class TryRegisterKeyedScopedAttribute : System.Attribute
+{ 
+    internal TryRegisterKeyedScopedAttribute(string ServiceKey, params System.Type[] onlyRegisterAs) { }
```
Comment (patrickklaeren) [inline src/AutoRegisterInject.Tests/GenerationTests.cs]:
```suggestion
    internal TryRegisterKeyedScopedAttribute(string serviceKey, params System.Type[] onlyRegisterAs) { }
```

File: src/AutoRegisterInject.Tests/GenerationTests.cs [hunk 5]
```diff
@@ -23,16 +23,73 @@ internal sealed class RegisterScopedAttribute : System.Attribute
...
 [System.AttributeUsage(System.AttributeTargets.Class, Inherited = false, AllowMultiple = false)]
 internal sealed class RegisterSingletonAttribute : System.Attribute
 { 
     internal RegisterSingletonAttribute(params System.Type[] onlyRegisterAs) { }
 }
+
+[System.AttributeUsage(System.AttributeTargets.Class, Inherited = false, AllowMultiple = false)]
+internal sealed class RegisterKeyedSingletonAttribute : System.Attribute
+{ 
+    internal RegisterKeyedSingletonAttribute(string ServiceKey, params System.Type[] onlyRegisterAs) { }
```
Comment (patrickklaeren) [inline src/AutoRegisterInject.Tests/GenerationTests.cs]:
```suggestion
    internal RegisterKeyedSingletonAttribute(string serviceKey, params System.Type[] onlyRegisterAs) { }
```

File: src/AutoRegisterInject.Tests/GenerationTests.cs [hunk 6]
```diff
@@ -23,16 +23,73 @@ internal sealed class RegisterScopedAttribute : System.Attribute
...
+[System.AttributeUsage(System.AttributeTargets.Class, Inherited = false, AllowMultiple = false)]
+internal sealed class TryRegisterSingletonAttribute : System.Attribute
+{ 
+    internal TryRegisterSingletonAttribute(params System.Type[] onlyRegisterAs) { }
+}
+
+[System.AttributeUsage(System.AttributeTargets.Class, Inherited = false, AllowMultiple = false)]
+internal sealed class TryRegisterKeyedSingletonAttribute : System.Attribute
+{ 
+    internal TryRegisterKeyedSingletonAttribute(string ServiceKey, params System.Type[] onlyRegisterAs) { }
```
Comment (patrickklaeren) [inline src/AutoRegisterInject.Tests/GenerationTests.cs]:
```suggestion
    internal TryRegisterKeyedSingletonAttribute(string serviceKey, params System.Type[] onlyRegisterAs) { }
```

File: src/AutoRegisterInject.Tests/GenerationTests.cs [hunk 7]
```diff
@@ -23,16 +23,73 @@ internal sealed class RegisterScopedAttribute : System.Attribute
...
 [System.AttributeUsage(System.AttributeTargets.Class, Inherited = false, AllowMultiple = false)]
 internal sealed class RegisterTransientAttribute : System.Attribute
 { 
     internal RegisterTransientAttribute(params System.Type[] onlyRegisterAs) { }
 }
+
+[System.AttributeUsage(System.AttributeTargets.Class, Inherited = false, AllowMultiple = false)]
+internal sealed class RegisterKeyedTransientAttribute : System.Attribute
+{ 
+    internal RegisterKeyedTransientAttribute(string ServiceKey, params System.Type[] onlyRegisterAs) { }
```
Comment (patrickklaeren) [inline src/AutoRegisterInject.Tests/GenerationTests.cs]:
```suggestion
    internal RegisterKeyedTransientAttribute(string serviceKey, params System.Type[] onlyRegisterAs) { }
```

File: src/AutoRegisterInject.Tests/GenerationTests.cs [hunk 8]
```diff
@@ -23,16 +23,73 @@ internal sealed class RegisterScopedAttribute : System.Attribute
...
+[System.AttributeUsage(System.AttributeTargets.Class, Inherited = false, AllowMultiple = false)]
+internal sealed class TryRegisterTransientAttribute : System.Attribute
+{ 
+    internal TryRegisterTransientAttribute(params System.Type[] onlyRegisterAs) { }
+}
+
+[System.AttributeUsage(System.AttributeTargets.Class, Inherited = false, AllowMultiple = false)]
+internal sealed class TryRegisterKeyedTransientAttribute : System.Attribute
+{ 
+    internal TryRegisterKeyedTransientAttribute(string ServiceKey, params System.Type[] onlyRegisterAs) { }
```
Comment (patrickklaeren) [inline src/AutoRegisterInject.Tests/GenerationTests.cs]:
```suggestion
    internal TryRegisterKeyedTransientAttribute(string serviceKey, params System.Type[] onlyRegisterAs) { }
```

**Review by patrickklaeren — CHANGES_REQUESTED** [review]

**Comment by patrickklaeren** [comment]
Minor change for the parameter name in constructor.

Will just need summary of attributes being added. Readme and Wiki needs to updated once released, I will add that as a chore for me to do.

**Comment by ChaseDRedmon** [comment]
Summary:

Added 3 new attributes types

1. TryRegister`XXX`
2. TryRegisterKeyed`XXX`
3. RegisterKeyed`XXX`

TryRegisterKeyed and RegisterKeyed attributes take a `string` `ServiceKey` argument where you must pass in a key. Failing to pass in a key, or setting an empty or whitespace key results in an argument exception that prevents the source generator from executing.

TryRegisterScoped, Singleton, and Transient attributes generate their corresponding extension methods: `ServiceCollection.TryAddScoped<>(), ServiceCollection.TryAddSingleton<>(), and ServiceCollection.TryAddTransient<>()`

`Try`RegisterKeyedScoped, KeyedSingleton, and KeyedTransient attributes generate their corresponding extension methods: `ServiceCollection.TryAddKeyedScoped<>("Key"), ServiceCollection.TryAddKeyedSingleton<>("Key"), and ServiceCollection.TryAddKeyedTransient<>("Key")`

RegisterKeyedScoped, KeyedSingleton, and KeyedTransient attributes generate their corresponding extension methods: `ServiceCollection.AddKeyedScoped<>("Key"), ServiceCollection.AddKeyedSingleton<>("Key"), and ServiceCollection.AddKeyedTransient<>("Key")`

**Pull request merged**